## Выпускной проект по теме SQL

Наша компания купила крупный сервис для чтения книг по подписке. 

### Задача - проанализировать базу данных.

В базе данных информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.



### Описание данных
1. Таблица 'books'

Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

2. Таблица 'authors'

Содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

3. аблица 'publishers'

Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;
4. Таблица 'ratings'

Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

5. Таблица 'reviews'

Содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

## Подключение БД

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
# посмотрим базы данных
query = ''' 
SELECT * 
FROM books
LIMIT 15'''
pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0         1        546                                       'Salem's Lot   
1         2        465                 1 000 Places to See Before You Die   
2         3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3         4         82  1491: New Revelations of the Americas Before C...   
4         5        125                                               1776   
5         6        257               1st to Die (Women's Murder Club  #1)   
6         7        258               2nd Chance (Women's Murder Club  #2)   
7         8        260              4th of July (Women's Murder Club  #4)   
8         9        563                                   A Beautiful Mind   
9        10        445                                 A Bend in the Road   
10       11        130         A Breath of Snow and Ashes (Outlander  #6)   
11       12        541                            A Brief History of Time   
12       13        359            A Caress of Twilight (Merry Gentry  #2)   
13       14         85     A Christmas Carol and Other Christmas Writings   
14       15        198     A Clash of Kings  (A Song of Ice and Fire  #2)   

    num_pages publication_date  publisher_id  
0         594       2005-11-01            93  
1         992       2003-05-22           336  
2         322       2010-12-21           135  
3         541       2006-10-10           309  
4         386       2006-07-04           268  
5         424       2005-05-20           116  
6         400       2005-05-20           116  
7         448       2006-06-01           318  
8         461       2002-02-04           104  
9         341       2005-04-01           116  
10        980       2006-08-29            83  
11        212       1998-09-01            36  
12        348       2005-03-15            33  
13        288       2003-10-30           217  
14        969       2002-05-28            35

In [3]:
query = ''' 
SELECT * 
FROM authors
LIMIT 15'''
pd.io.sql.read_sql(query, con = engine) 

author_id                                author
0           1                            A.S. Byatt
1           2        Aesop/Laura Harris/Laura Gibbs
2           3                       Agatha Christie
3           4                         Alan Brennert
4           5              Alan Moore/David   Lloyd
5           6                            Alan Paton
6           7           Albert Camus/Justin O'Brien
7           8                         Aldous Huxley
8           9    Aldous Huxley/Christopher Hitchens
9          10  Aleksandr Solzhenitsyn/H.T. Willetts
10         11                Alexander McCall Smith
11         12      Alexander Pushkin/James E. Falen
12         13            Alexandre Dumas/Robin Buss
13         14                         Alice Hoffman
14         15                          Alice Walker

In [4]:
query = ''' 
SELECT * 
FROM publishers
LIMIT 15'''
pd.io.sql.read_sql(query, con = engine) 

publisher_id                                publisher
0              1                                      Ace
1              2                                 Ace Book
2              3                                Ace Books
3              4                            Ace Hardcover
4              5        Addison Wesley Publishing Company
5              6                                  Aladdin
6              7                       Aladdin Paperbacks
7              8                             Albin Michel
8              9                          Alfred A. Knopf
9             10  Alfred A. Knopf Books for Young Readers
10            11             Alfred A. Knopf Borzoi Books
11            12                          Algonquin Books
12            13                            Allen & Ulwin
13            14                                  Amistad
14            15                                   Anchor

In [5]:
query = ''' 
SELECT * 
FROM ratings
LIMIT 15'''
pd.io.sql.read_sql(query, con = engine) 

rating_id  book_id       username  rating
0           1        1     ryanfranco       4
1           2        1  grantpatricia       2
2           3        1   brandtandrea       5
3           4        2       lorichen       3
4           5        2    mariokeller       2
5           6        3  johnsonamanda       4
6           7        3    scotttamara       5
7           8        3    lesliegibbs       5
8           9        4    abbottjames       5
9          10        4   valenciaanne       4
10         11        5    jeromebowen       4
11         12        5        staylor       4
12         13        5        susan85       3
13         14        5        npowers       4
14         15        5        znelson       5

In [6]:
query = ''' 
SELECT * 
FROM reviews
LIMIT 15'''
pd.io.sql.read_sql(query, con = engine) 

review_id  book_id       username  \
0           1        1   brandtandrea   
1           2        1     ryanfranco   
2           3        2       lorichen   
3           4        3  johnsonamanda   
4           5        3    scotttamara   
5           6        3    lesliegibbs   
6           7        4   valenciaanne   
7           8        4    abbottjames   
8           9        5        npowers   
9          10        5        staylor   
10         11        5       thomas49   
11         12        5        znelson   
12         13        6       walter49   
13         14        6   jasongoodman   
14         15        6         eallen   

                                                 text  
0   Mention society tell send professor analysis. ...  
1   Foot glass pretty audience hit themselves. Amo...  
2   Listen treat keep worry. Miss husband tax but ...  
3   Finally month interesting blue could nature cu...  
4   Nation purpose heavy give wait song will. List...  
5            Analysis no several cause international.  
6   One there cost another. Say type save. With pe...  
7   Within enough mother. There at system full rec...  
8   Thank now focus realize economy focus fly. Ite...  
9   Game push lot reduce where remember. Including...  
10  Their physical over rise. Couple doctor seven ...  
11  Explain individual record wonder we must turn ...  
12  Western anyone until executive. Add teach ago ...  
13  Much responsibility feel fund media. Wind musi...  
14                             Money arm first above.

## Задания

### Посчитаем, сколько книг вышло после 1 января 2000 года

In [7]:
query = '''
SELECT COUNT(DISTINCT book_id) AS count_new_books
    FROM
        books
    WHERE
        publication_date > '2000-01-01' '''
    
    
    
pd.io.sql.read_sql(query, con = engine)

count_new_books
0              819

### Для каждой книги посчитаем количество обзоров и среднюю оценку

In [8]:
query = '''
SELECT 
    books.title AS name_book,
    COUNT(DISTINCT reviews.review_id) AS count_reviews,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    INNER JOIN reviews ON reviews.book_id=books.book_id
    INNER JOIN ratings ON ratings.book_id=books.book_id
GROUP BY
    books.title 
ORDER BY 
    count_reviews DESC
LIMIT 10 '''
    
pd.io.sql.read_sql(query, con = engine)

name_book  count_reviews  \
0                                Memoirs of a Geisha              8   
1                            Twilight (Twilight  #1)              7   
2  The Curious Incident of the Dog in the Night-Time              6   
3                          Outlander (Outlander  #1)              6   
4                                     The Book Thief              6   
5                             The Catcher in the Rye              6   
6  Harry Potter and the Chamber of Secrets (Harry...              6   
7                                    Eat  Pray  Love              6   
8  Harry Potter and the Prisoner of Azkaban (Harr...              6   
9                                      The Alchemist              6   

   avg_rating  
0    4.138462  
1    3.662500  
2    4.081081  
3    4.125000  
4    4.264151  
5    3.825581  
6    4.287500  
7    3.395833  
8    4.414634  
9    3.789474

In [9]:
query = '''
SELECT 
    books.title AS name_book,
    COUNT(DISTINCT reviews.review_id) AS count_reviews,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    INNER JOIN reviews ON reviews.book_id=books.book_id
    INNER JOIN ratings ON ratings.book_id=books.book_id
GROUP BY
    books.title
ORDER BY 
    avg_rating DESC
LIMIT 50 '''
  
pd.io.sql.read_sql(query, con = engine)

name_book  count_reviews  \
0                   Misty of Chincoteague (Misty  #1)              2   
1           A Woman of Substance (Emma Harte Saga #1)              2   
2   The War of Art: Break Through the Blocks & Win...              2   
3                                       Evening Class              2   
4   Wherever You Go  There You Are: Mindfulness Me...              2   
5                                          Dead Souls              2   
6                            Tai-Pan (Asian Saga  #2)              2   
7   The Ghost Map: The Story of London's Most Terr...              2   
8                                         Angels Fall              2   
9                                       Alas  Babylon              2   
10                          Triptych (Will Trent  #1)              2   
11                                 The Woman in Black              2   
12               Welcome to Temptation (Dempseys  #1)              2   
13  Stone of Farewell (Memory  Sorrow  and Thorn  #2)              1   
14                                    Light in August              2   
15                           Neil Gaiman's Neverwhere              2   
16  The Demon-Haunted World: Science as a Candle i...              2   
17                 The Big Bad Wolf  (Alex Cross  #9)              2   
18                                        Marvel 1602              2   
19                                              Women              2   
20                               My Name Is Asher Lev              2   
21       Moneyball: The Art of Winning an Unfair Game              3   
22               Piercing the Darkness (Darkness  #2)              2   
23               Pop Goes the Weasel (Alex Cross  #5)              2   
24                       Plum Island (John Corey  #1)              2   
25                      Emily of New Moon (Emily  #1)              2   
26  The Adventures of Tom Sawyer and Adventures of...              1   
27            School's Out—Forever (Maximum Ride  #2)              3   
28           Evil Under the Sun (Hercule Poirot  #24)              2   
29               Stolen (Women of the Otherworld  #2)              2   
30                                Homage to Catalonia              2   
31                                          Geek Love              2   
32                      A Dirty Job (Grim Reaper  #1)              4   
33                                         Hard Times              2   
34  In the Hand of the Goddess (Song of the Liones...              2   
35                    Act of Treason (Mitch Rapp  #9)              2   
36      Arrows of the Queen (Heralds of Valdemar  #1)              2   
37  How to Be a Domestic Goddess: Baking and the A...              1   
38              A Fistful of Charms (The Hollows  #4)              2   
39  Crucial Conversations: Tools for Talking When ...              2   
40                                              March              2   
41  The Walking Dead  Book One (The Walking Dead #...              2   
42  Captivating: Unveiling the Mystery of a Woman'...              2   
43                 Dragonfly in Amber (Outlander  #2)              4   
44             Preludes & Nocturnes (The Sandman  #1)              3   
45  The Witching Hour (Lives of the Mayfair Witche...              3   
46      Lover Revealed (Black Dagger Brotherhood  #4)              3   
47                The Monster at the End of this Book              3   
48                                          The Twits              3   
49                                    House of Leaves              4   

    avg_rating  
0     5.000000  
1     5.000000  
2     5.000000  
3     5.000000  
4     5.000000  
5     5.000000  
6     5.000000  
7     5.000000  
8     5.000000  
9     5.000000  
10    5.000000  
11    5.000000  
12    5.000000  
13    5.000000  
14    5.000000  
15    5.000000  
16    5.000000  
17    5.000000  
18    5.000000  
19    5.000000  
20    5.000000

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц

In [10]:
query = '''
SELECT 
    publishers.publisher AS name_publisher,
    COUNT(books.num_pages > 50) AS counts
FROM 
    books
    INNER JOIN publishers ON publishers.publisher_id=books.publisher_id
GROUP BY
    publishers.publisher
ORDER BY 
    counts DESC
LIMIT 1 '''
    
pd.io.sql.read_sql(query, con = engine)

name_publisher  counts
0  Penguin Books      42

### Определим автора с самой высокой средней оценкой книг (более 50 оценок)

In [11]:
query = '''
SELECT 
    authors.author,
    authors.author_id,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    INNER JOIN ratings ON ratings.book_id=books.book_id
    INNER JOIN authors ON authors.author_id=books.author_id
WHERE
    authors.author_id IN (SELECT
                              books.author_id
                         FROM
                             books
                         WHERE
                         books.book_id IN (SELECT 
                                                ratings.book_id
                                           FROM
                                               ratings
                                           GROUP BY
                                               ratings.book_id
                                           HAVING
                                               COUNT(ratings.rating_id) > 50))
GROUP BY
    authors.author, authors.author_id
ORDER BY 
    avg_rating DESC 
LIMIT 1'''
    
pd.io.sql.read_sql(query, con = engine)

author  author_id  avg_rating
0  J.K. Rowling/Mary GrandPré        236    4.288462

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [12]:
query = '''
SELECT 
    SUM(SUBQ.cnt_reviews) / COUNT(SUBQ.cnt_reviews) AS cnt
FROM (
    SELECT
        reviews.username,
        COUNT(reviews.username) AS cnt_reviews
    FROM
        reviews
    WHERE
        reviews.username IN (SELECT
                                 ratings.username
                             FROM 
                                 ratings
                             GROUP BY
                                 ratings.username
                             HAVING
                                 COUNT(ratings.rating_id) > 50
                                 
                             )
    GROUP BY
        username) AS SUBQ '''

pd.io.sql.read_sql(query, con = engine)

cnt
0  24.333333

## Вывод

Мы подключились к базе данных и посмотрели на имеющуюся информацию

Решили задания и выяснили:
- После 1 января 2000 года вышло 819 книг
- Вывели на экран количество обзоров и среднюю оценку для каждой книги
- Издательстов, выпустившее больше всего книг - Penguin Books
- Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.3